packages/clay-color-picker/src/hooks.ts

```diff
@@ -0,0 +1,86 @@
+/**
+ * © 2019 Liferay, Inc. <https://liferay.com>
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+import * as React from 'react';
+
+/**
+ * Utility hook for calculating the mouse position
+ */
+export function useMousePosition(containerRef: React.RefObject<any>) {
+	const [xy, setXY] = React.useState({x: 0, y: 0});
+
+	const onMouseMove = (event: React.MouseEvent | MouseEvent) => {
+		const rect = containerRef.current.getBoundingClientRect();
+
+		const x = event.pageX;
+
+		let newLeft = x - (rect.left + window.pageXOffset);
+
+		newLeft = newLeft < 0 ? 0 : newLeft > rect.width ? rect.width : newLeft;
+
+		const y = event.pageY;
+
+		let newTop = y - (rect.top + window.pageYOffset);
+
+		newTop = newTop < 0 ? 0 : newTop > rect.height ? rect.height : newTop;
+
+		setXY({x: newLeft, y: newTop});
+	};
+
+	return {...xy, onMouseMove, setXY};
+}
+
+/**
+ * Regex to only handle values of A,B,C,D,E,F,0,1,2,3,4,5,6,7,8,9
+ */
+const HEX_REGEX = /^[a-fA-F0-9]+$/;
+
+/**
+ * Hook for updating input value
+ */
+export function useHexInput(hex: string): [string, (val: string) => void] {
+	const [inputValue, setInputValue] = React.useState(hex);
+
+	const handleNewInputValue = (value: string) => {
+		const match = value.match(HEX_REGEX);
+
+		setInputValue(match ? match[0] : '');
+	};
+
+	return [inputValue, handleNewInputValue];
+}
+
+const ESC_KEY_CODE = 27;
+
+/**
+ * Hook for closing dropdown when user hits ESC key or clicks outside the menu.
+ */
+export function useDropdownCloseInteractions(
```

Comment:
hey @bryceosterhaus I've been thinking of being able to share this rule with the other components, we have Dropdown here and DatePicker, both have the same behavior. But I wish it were not a dependency of a single component and that we have to import from it to use and also does not have to be a package of utilitarios of Clay, I do not want that we commit to him to make available, only for internal use that we can modify whenever I needed it, I've been thinking of something like the `shared` folder something like React does internally in its codebase. What do you think?

Reply:
Yeah thats a great idea. I think we could get away with either a generic `@clayui/shared` package or a hooks specific package `@clayui/shared-hooks`

Reply:
Before we extract it though, I think its best to keep this here and then once dropdown is created, we can reconcile the two and move it to shared.

Reply:
I think we can only have a `shared` to add components (which are just small rules), utils and hooks if necessary, but I did not want it to be handled as a package, if doing that we would have to publish, now I wanted to avoid if that.

> Before we extract it though, I think its best to keep this here and then once dropdown is created, we can reconcile the two and move it to shared.

Yes No problems.

Reply:
Oh gotcha, i didn't see your reference to not publishing it. Yeah I think that is fine for now, not sure if we can avoid publishing it in the future though since it will be controlled by semver somewhere.

Reply:
Yes I thought about it, maybe we can handle it at build time so it does not have to be published. The concern I have in publishing it will be another package that will be available for people to use, so we would have to follow it as if it were another component keeping the semver and the good maintenance practices of not breaking, I wish we had the freedom to change those files without having to worry about it.